sima/motion/dftreg.py

```diff
@@ -0,0 +1,788 @@
+2015 Lloyd Russell, Christoph Schmidt-Hieber
+
+*******************************************************************************
+Credit to Marius Pachitariu for concept of registering to aligned mean image.
+
+Parts of the code are based on:
+skimage.feature.register_translation, which is a port of MATLAB code by Manuel
+Guizar-Sicairos, Samuel T. Thurman, and James R. Fienup, "Efficient subpixel
+image registration algorithms," Optics Letters 33, 156-158 (2008).
+
+Relating to implementation of skimage.feature.register_translation:
+Copyright (C) 2011, the scikit-image team
+All rights reserved.
+THIS SOFTWARE IS PROVIDED BY THE AUTHOR ''AS IS'' AND ANY EXPRESS OR IMPLIED
+WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO
+EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
+OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
+IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY
+OF SUCH DAMAGE.
+*******************************************************************************
+@author: llerussell
+"""
+
+from __future__ import absolute_import, division
+from builtins import map, range
+from functools import partial
+import multiprocessing
+import numpy as np
+from scipy.ndimage.interpolation import shift
+import time
+from . import motion
+
+
+class DiscreteFourier2D(motion.MotionEstimationStrategy):
+    """
+    Motion correction of image sequences by 'efficient subpixel image
+    registration by cross correlation'. A reference image is iteratively
+    computed by aligning and averaging a subset of images/frames.
+
+    Parameters
+    ----------
+    upsample_factor : int
+        upsample factor. final pixel alignment has resolution of
+        1/upsample_factor. if 1 only pixel level shifts are made - faster -
+        and no interpolation (optional, default=1)
+    max_displacement : array of int
+        The maximum allowed displacement magnitudes in [y,x] (optional,
+        default = None)
+    num_images_for_mean : int
+        number of images to use to make the aligned mean image (optional,
+        default=100)
+    randomise_frames : bool
+        randomise the images selected to make the mean image? if false the
+        first 'num_frames_for_mean' frames will be used (optional,
+        default=True)
+    err_thresh : float
+        the error threshold level at which to stop iterating over the mean
+        image alignment (optional, default=0.01)
+    max_iterations : int
+        the maximum number of iterations to compute the aligned mean image
+        (optional, default=5)
+    use_fftw : bool
```

Comment:
Rather than having an argument for this, I would just have the code use ```fftw``` whenever it is available.

Here is how we did it in sima/misc/align.py:

```python
try:
    from pyfftw.interfaces.scipy_fftpack import fft2, ifft2
    from pyfftw.interfaces.numpy_fft import rfftn, irfftn
except ImportError:
    from scipy.fftpack import fft2, ifft2
    from numpy.fft import rfftn, irfftn
```